base/strings/util.jl

```diff
@@ -689,17 +689,19 @@ julia> bytes2hex(b)
 """
 function bytes2hex end
 
-function bytes2hex(a::Union{Tuple{Vararg{UInt8}}, AbstractArray{UInt8}})
-    b = Base.StringVector(2*length(a))
-    @inbounds for (i, x) in enumerate(a)
+function bytes2hex(itr)
+    eltype(itr) != UInt8 && throw(ArgumentError("eltype of iterator not UInt8"))
```

Comment:
Unsure whether I should use `!==` here instead.

Reply:
`!==` would be better, but I prefer `eltype(itr) === UInt8 ||`.

Reply:
I think this check is not even necessary --- the type of each `x` could be checked instead.

Reply:
In the case of `bytes2hex(itr)`, that's fine (if the type returned by the iterator is known, branch elimination could be done, I think).

However, in the case of `bytes2hex(io, itr)`, it will pollute `io` until the problematic element appears.

Reply:
If the type is always going to be UInt8, the branch is indeed eliminated, but as soon as you put in e.g. Vector{Union{UInt8,Int}}, performance becomes abysmal.

```
julia> function b2h(itr)
         b = Base.StringVector(2*length(itr))
         @inbounds for (i,x) in enumerate(itr)
           typeof(x) === UInt8 || throw(ArgumentError("iterator has non-UInt8 elements!"))
           b[2i - 1] = Base.hex_chars[1 + x >> 4]
           b[2i    ] = Base.hex_chars[1 + x & 0xf]
         end
         return String(b)
       end
b2h (generic function with 1 method)

julia> @benchmark b2h(str) setup=(str=Vector{Union{UInt8,Int}}(collect(transcode(UInt8, randstring("0123456789abcdef", 1_000_000)))))
BenchmarkTools.Trial:
  memory estimate:  1.91 MiB
  allocs estimate:  2
  --------------
  minimum time:     9.836 ms (0.00% GC)
  median time:      10.405 ms (0.00% GC)
  mean time:        10.450 ms (0.00% GC)
  maximum time:     11.663 ms (0.00% GC)
  --------------
  samples:          226
  evals/sample:     1

julia> @benchmark b2h(str) setup=(str=collect(transcode(UInt8, randstring("0123456789abcdef", 1_000_000))))
BenchmarkTools.Trial:
  memory estimate:  1.91 MiB
  allocs estimate:  2
  --------------
  minimum time:     857.800 μs (0.00% GC)
  median time:      906.200 μs (0.00% GC)
  mean time:        922.933 μs (0.00% GC)
  maximum time:     1.762 ms (0.00% GC)
  --------------
  samples:          507
  evals/sample:     1
```

Not to mention that, should we encounter a non-UInt8 element, we'd have to throw either way. We could put some other function in front of the `AbstractArray{UInt8}` version to automatically convert everything to its byte representation. We probably only want to do this in the general discussion about reworking the n2x interface I've seen in #14418.

Reply:
The performance hit for the IO version is not as bad, but that's mostly due to the fact that this is already pretty slow compared to not printing to an IO:

```
julia> function b2h(io::IO, itr)
           for x in itr
               typeof(x) === UInt8 || throw(ArgumentError("element not UInt8"))
               print(io, Char(Base.hex_chars[1 + x >> 4]), Char(Base.hex_chars[1 + x & 0xf]))
           end
       end
b2h (generic function with 2 methods)
julia> @benchmark b2h(io,str) setup=(str=rand(transcode(UInt8, "0123456789abcdef"), 1_000_000);io=IOBuffer())
BenchmarkTools.Trial:
  memory estimate:  2.00 MiB
  allocs estimate:  16
  --------------
  minimum time:     44.196 ms (0.00% GC)
  median time:      45.609 ms (0.00% GC)
  mean time:        45.811 ms (0.00% GC)
  maximum time:     50.710 ms (0.00% GC)
  --------------
  samples:          98
  evals/sample:     1

julia> @benchmark b2h(io,str) setup=(str=Vector{Union{UInt8,UInt}}(rand(transcode(UInt8, "0123456789abcdef"), 1_000_000));io=IOBuffer())
BenchmarkTools.Trial:
  memory estimate:  2.00 MiB
  allocs estimate:  16
  --------------
  minimum time:     50.614 ms (0.00% GC)
  median time:      51.852 ms (0.00% GC)
  mean time:        52.087 ms (0.00% GC)
  maximum time:     56.459 ms (0.00% GC)
  --------------
  samples:          84
  evals/sample:     1
```